### DISCUSSION TITLE
Build failed: Deprecated Gradle features

### DISCUSSION THREAD

**Discussion by fuadreza** [codemagic-ci-cd/codemagic-docs#1367] [Q&A] [answered]
Hello,

I've been encountered flutter build error using code magic. I run the build with command:
```
flutter build appbundle --release --build-name=$APP_VERSION --build-number=$LATEST_GOOGLE_PLAY_BUILD_NUMBER
```

and stop with this error

```
[        ] BUILD FAILED in 1m 50s
[        ] Deprecated Gradle features were used in this build, making it incompatible with Gradle 8.0.
[        ] You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.
[        ] See https://docs.gradle.org/7.3.3/userguide/command_line_interface.html#sec:command_line_warnings
[        ] 204 actionable tasks: 204 executed
[        ] Watched directory hierarchies: []
[ +683 ms] Running Gradle task 'bundleRelease'... (completed in 112.7s)
[   +7 ms] "flutter appbundle" took 113,087ms.
[   +6 ms] Gradle task bundleRelease failed with exit code 1
```

But if I build that on my local machine, it works fine. I also check the Gradle using command:
```
./gradlew --warning-mode all
```
and return something like this:
```
> Configure project :share_plus
Subproject ':share_plus' has location '/Users/fuadrezapahlevi/Documents/Development/Flutter/flutter/.pub-cache/hosted/pub.dartlang.org/share_plus-6.1.0/android' which is outside of the project root. This behaviour has been deprecated and is scheduled to be removed in Gradle 8.0. Consult the upgrading guide for further information: https://docs.gradle.org/7.3.3/userguide/upgrading_version_7.html#deprecated_flat_project_structure

> Configure project :url_launcher_android
Subproject ':url_launcher_android' has location '/Users/fuadrezapahlevi/Documents/Development/Flutter/flutter/.pub-cache/hosted/pub.dartlang.org/url_launcher_android-6.0.17/android' which is outside of the project root. This behaviour has been deprecated and is scheduled to be removed in Gradle 8.0. Consult the upgrading guide for further information: https://docs.gradle.org/7.3.3/userguide/upgrading_version_7.html#deprecated_flat_project_structure
```

How am I supposed to do with this? Thank you..

**Comment by himesh-cm** [comment]
Hi @fuadreza, Can you please share the build id so I can further look into your query?

**Reply by fuadreza** [reply]
Hi @himesh-cm , where can I find my build id? Thankyou

**Reply by fuadreza** [reply]
63d897d93f44bc12f0d57810 is it this?

**Reply by himesh-cm** [reply]
Yes, I will check and get back to you.

**Reply by himesh-cm** [reply]
Hi @fuadreza,
Configuration-wise, everything looks ok; for debugging purposes, can you run the last successful commit, `e7eaf3e` again?

Error, but it isn't easy to point out the root cause.
```
FAILURE: Build failed with an exception.
[        ] * Where:
[        ] Script '/Users/builder/programs/flutter/packages/flutter_tools/gradle/flutter.gradle' line: 1159
[        ] * What went wrong:
[        ] Execution failed for task ':app:compileFlutterBuildRelease'.
[        ] > Process 'command '/Users/builder/programs/flutter/bin/flutter'' finished with non-zero exit value 1
```

**Reply by fuadreza** [reply]
Hi @himesh-cm, 
I tried to build with `e7eaf3e` commit as you mentioned, but it turned out gives me error :

```
[        ] BUILD FAILED in 1m 47s
[        ] Deprecated Gradle features were used in this build, making it incompatible with Gradle 8.0.
[        ] You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.
[        ] See https://docs.gradle.org/7.3.3/userguide/command_line_interface.html#sec:command_line_warnings
```

You can check in this id `63d8cf8ddd8b47b6aa2bd999`

I'm aware if there is some changes in gradle version in the virtual machine, but I'm not really sure.

**Reply by himesh-cm** [reply]
Hi @fuadreza , Looks like a known issue as seen [here](https://github.com/flutter/flutter/issues/119444)
Let me check internally with team, just to confirm

**Reply by himesh-cm** [reply]
Hi @fuadreza,
It seems the error is related to
```
[+5699 ms] ../programs/flutter/.pub-cache/hosted/pub.dartlang.org/flutter_typeahead-4.3.1/lib/src/flutter_typeahead.dart:898:40: Error: Member not found: 'Scrollable.maybeOf'.
[        ]     final scrollableState = Scrollable.maybeOf(context);
[        ]                                        ^^^^^^^
[        ] ../programs/flutter/.pub-cache/hosted/pub.dartlang.org/flutter_typeahead-4.3.1/lib/src/cupertino_flutter_typeahead.dart:649:40: Error: Member not found: 'Scrollable.maybeOf'.
[        ]     final scrollableState = Scrollable.maybeOf(context);
```
more info here
```
package: https://pub.dev/packages/flutter_typeahead/changelog
Related issue: https://github.com/AbdulRahmanAlHamali/flutter_typeahead/issues/449
```

Can you please upgrade to flutter `3.7.0` and try rerunning the build?

**Reply by himesh-cm** [reply]
Or you can lower the version as mentioned [here](https://github.com/AbdulRahmanAlHamali/flutter_typeahead/issues/449#issuecomment-1408591182), if you want to stick with current flutter version used in the project.

**Reply by fuadreza** [reply]
Hi @himesh-cm ,
I'm sorry I don't understand why is it the case.. I mean, I run the build for Android not for ios as you mentioned in the issue above.

Also I already set my flutter version to `3.3.4` inside my `codemagic.yaml`. Here is my snippet of my `codemagic.yaml` configuration:
<img width="158" alt="image" src="https://user-images.githubusercontent.com/25898509/215761226-234a8c54-a816-404a-b0b8-e8ddb264a59b.png">

Is there any method I can try other than that? Thanks in advance.

**Reply by fuadreza** [reply]
Hi @himesh-cm ,

Any update on this? Thank you.

**Reply by fuadreza** [reply]
Ok, after I upgrade my flutter and fix deprecated code and now the build success.

I'm just thinking if we can still use the old version of flutter to make the build run.

**Reply by fuadpahlevi-jubelio** [reply]
Hi @himesh-cm, is there anything I can do to lock flutter version? I don't want to upgrade my flutter version at the moment. 

Thanks

**Reply by dtrdic** [reply]
hey @fuadreza there is a Flutter Version Management (FVM) tool available that allows you to easily select the desired Flutter version specified in your FVM configuration file. You can download the FVM package from here https://fvm.app/docs/getting_started/overview/

This will ensure that the build machine will always use the Flutter version specified in your FVM configuration.

Check more about setting flutter version [HERE](https://docs.codemagic.io/yaml-quick-start/building-a-flutter-app/#setting-the-flutter-version)

**Reply by fuadreza** [reply]
@dtrdic Thanks for the answer. But I manage to lock the flutter version in codemagic.yaml

**Reply by dtrdic** [reply]
Gj, @fuadreza i'm glad that it works for you:) can you share what modifications have you done so we can mark it as the answer for other users?

**Comment by fuadreza** [comment, marked as answer]
Hi @dtrdic,
After further checking, my problem happen because some of my library is using deprecated `jcenter()` inside android `build.gradle` file. Somehow this makes codemagic to fails when building the android app. 

What you can do to check is to run `./gradlew --warning-mode all` inside `android` folder in your flutter project

This is some example of the warning in deprecated:
```
The RepositoryHandler.jcenter() method has been deprecated. This is scheduled to be removed in Gradle 8.0. JFrog announced JCenter's sunset in February 2021. 
Use mavenCentral() instead. Consult the upgrading guide for further information: https://docs.gradle.org/7.3.3/userguide/upgrading_version_6.html#jcenter_deprecation
        at build_5gkwqxn2cv7n6v2imlk5mynyc$_run_closure1$_closure2.doCall(/Users/xxx/.pub-cache/git/presentation-displays-b6bbe5c7163679298340920f793c72814c40d611/android/build.gradle:8)
        (Run with --stacktrace to get the full stack trace of this deprecation warning.)

```

What I do is:
Fork that library and then change `jcenter()` to `mavenCentral()`. Change the library version in `pubspec.yaml` to use my forked version of the library.
